### PR TITLE
refactored GaussianKernel as a subclass of CShiftInvariantKernel

### PR DESCRIPTION
**Pull request by lambday**
@karlnapf please have a  look!

### CODE REVIEW

File: src/shogun/kernel/GaussianKernel.cpp [hunk 1]
```diff
@@ -63,13 +53,9 @@ CGaussianKernel::~CGaussianKernel()
 
 CGaussianKernel* CGaussianKernel::obtain_from_generic(CKernel* kernel)
 {
-	if (kernel->get_kernel_type()!=K_GAUSSIAN)
-	{
-		SG_SERROR("CGaussianKernel::obtain_from_generic(): provided kernel is "
-				"not of type CGaussianKernel!\n");
-	}
+	REQUIRE(kernel->get_kernel_type()==K_GAUSSIAN,
+		"Provided kernel (%s) must be of type CGaussianKernel!\n", kernel->get_name());
```
Comment (karlnapf) [inline src/shogun/kernel/GaussianKernel.cpp]: maybe use self->name here rather than CGaussianKernel?

File: src/shogun/kernel/GaussianKernel.cpp [hunk 2]
```diff
@@ -82,86 +68,55 @@ CSGObject *CGaussianKernel::shallow_copy() const
...
-	REQUIRE(lhs && rhs, "Features not set!\n")
+	REQUIRE(lhs, "Left hand side features must be set!\n")
+	REQUIRE(rhs, "Rightt hand side features must be set!\n")
 
 	if (!strcmp(param->m_name, "log_width"))
 	{
 		SGMatrix<float64_t> derivative=SGMatrix<float64_t>(num_lhs, num_rhs);
-
 		for (int j=0; j<num_lhs; j++)
+		{
```
Comment (karlnapf) [inline src/shogun/kernel/GaussianKernel.cpp]: openmp maybe?
But this is not the point here I guess

File: tests/unit/base/SGObject_unittest.cc [hunk 3]
```diff
@@ -208,7 +208,7 @@ TEST(SGObject,equals_complex_equal)
 	SG_UNREF(file);
 
 	/* now compare */
-	floatmax_t accuracy=1E-15;
+	floatmax_t accuracy=1E-10;
```
Comment (karlnapf) [inline tests/unit/base/SGObject_unittest.cc]: ?

**Comment by karlnapf** [comment]
Fine to merge from my side once trabis is green

File: src/shogun/kernel/GaussianKernel.cpp (hunk 1 above)
Comment (lambday) [inline src/shogun/kernel/GaussianKernel.cpp]: hehe tried that.. then realized that this is a static method :D

File: src/shogun/kernel/GaussianKernel.cpp [hunk 4]
```diff
@@ -82,86 +68,55 @@ CSGObject *CGaussianKernel::shallow_copy() const
...
 			for (int k=0; k<num_rhs; k++)
 			{
-				float64_t element=distance(j,k);
-				derivative(j,k)=exp(-element)*element*2.0;
+				float64_t element=distance(j, k);
```
Comment (lambday) [inline src/shogun/kernel/GaussianKernel.cpp]: actually this is slow if precomputed distance matrix is provided. maybe I'll try changing the order in a separate patch (making sure that col-wise traversal is actually beneficial)

File: src/shogun/kernel/GaussianKernel.cpp (hunk 2 above)
Comment (lambday) [inline src/shogun/kernel/GaussianKernel.cpp]: agreed

File: src/shogun/kernel/GaussianKernel.cpp [hunk 5]
```diff
@@ -82,86 +68,55 @@ CSGObject *CGaussianKernel::shallow_copy() const
...
-	if (lhs==rhs)
-		sq_rhs=sq_lhs;
-	else
-		precompute_squared_helper(sq_rhs, (CDotFeatures*) rhs);
+	return CMath::exp(m_log_width*2.0)*2.0;
 }
 
-SGMatrix<float64_t> CGaussianKernel::get_parameter_gradient(
-		const TParameter* param, index_t index)
+SGMatrix<float64_t> CGaussianKernel::get_parameter_gradient(const TParameter* param, index_t index)
```
Comment (vigsterkr) [inline src/shogun/kernel/GaussianKernel.cpp]: why does it only can be a `float64_t`? and not a `float32_t` or a `floatmax_t`?

File: src/shogun/kernel/GaussianKernel.cpp [hunk 6]
```diff
@@ -82,86 +68,55 @@ CSGObject *CGaussianKernel::shallow_copy() const
...
 		SGMatrix<float64_t> derivative=SGMatrix<float64_t>(num_lhs, num_rhs);
-
 		for (int j=0; j<num_lhs; j++)
+		{
 			for (int k=0; k<num_rhs; k++)
 			{
-				float64_t element=distance(j,k);
-				derivative(j,k)=exp(-element)*element*2.0;
+				float64_t element=distance(j, k);
+				derivative(j, k)=exp(-element)*element*2.0;
```
Comment (vigsterkr) [inline src/shogun/kernel/GaussianKernel.cpp]: for fun and profit: http://stackoverflow.com/questions/10552280/fast-exp-calculation-possible-to-improve-accuracy-without-losing-too-much-perfo/10792513#10792513

File: src/shogun/kernel/GaussianKernel.cpp [hunk 7]
```diff
@@ -171,15 +126,33 @@ SGMatrix<float64_t> CGaussianKernel::get_parameter_gradient(
...
 {
-	return (sq_lhs[idx_a]+sq_rhs[idx_b]-2*CDotKernel::compute(idx_a,idx_b))/get_width();
+	CKernel::load_serializable_post();
+	if (lhs && rhs)
+		m_distance->init(lhs, rhs);
+}
+
+float64_t CGaussianKernel::distance(int32_t idx_a, int32_t idx_b) const
+{
+	return CShiftInvariantKernel::distance(idx_a, idx_b)/get_width();
```
Comment (vigsterkr) [inline src/shogun/kernel/GaussianKernel.cpp]: mmm will clang/gcc in `-O3` optimize this into `1/get_width()` being a const and just do a `CShiftInvariantKernel::distance(idx_a, idx_b)*<that const>`?

File: src/shogun/kernel/GaussianKernel.cpp [hunk 8]
```diff
@@ -82,86 +68,55 @@ CSGObject *CGaussianKernel::shallow_copy() const
...
-	if (!lhs || !rhs)
-		return;
-
-	precompute_squared_helper(sq_lhs, (CDotFeatures*) lhs);
-
-	if (lhs==rhs)
-		sq_rhs=sq_lhs;
-	else
-		precompute_squared_helper(sq_rhs, (CDotFeatures*) rhs);
+	return CMath::exp(m_log_width*2.0)*2.0;
```
Comment (vigsterkr) [inline src/shogun/kernel/GaussianKernel.cpp]: see below about fastexp

File: src/shogun/kernel/GaussianKernel.cpp (hunk 5 above)
Comment (lambday) [inline src/shogun/kernel/GaussianKernel.cpp]: good point. it is coming from the fact that the distance we calculate is always `float64_t` no matter what. we should change that later. me and @karlnapf had a discussion on similar things a few days back. how about (just like in CKernel), we have a templated version of `CDistance::distance`, and then a non-templated version that (by default) returns float64_t.

File: src/shogun/kernel/GaussianKernel.cpp (hunk 7 above)
Comment (lambday) [inline src/shogun/kernel/GaussianKernel.cpp]: good point. not sure whether it does that. better to be explicit.

**Comment by lambday** [comment]
updated the PR with a few suggested changes. travis is green. merging this one.

File: src/shogun/kernel/GaussianKernel.cpp (hunk 5 above)
Comment (karlnapf) [inline src/shogun/kernel/GaussianKernel.cpp]: Yeah this is not good. But not the point of this PR and has been there before.
This patch is just some mini changes that we need for research code.
chris can deal with this, he is almost done with LARS

**Pull request merged**